src/pke/lib/scheme/ckksrns/ckksrns-schemeswitching.cpp

```diff
@@ -0,0 +1,2152 @@
+    auto oldTranformedSK = ccCKKSto->KeyGen().secretKey;
+    oldTranformedSK->SetPrivateElement(std::move(skElements));
+
+    return ccCKKSto->KeySwitchGen(oldTranformedSK, ckksSKto);
+}
+
+EvalKey<DCRTPoly> switchingKeyGenRLWEcc(const PrivateKey<DCRTPoly>& ckksSKto, const PrivateKey<DCRTPoly>& ckksSKfrom,
+                                        ConstLWEPrivateKey& LWEsk) {
+    auto skElements = ckksSKto->GetPrivateElement();
+    skElements.SetFormat(Format::COEFFICIENT);
+    auto skElementsFrom = ckksSKfrom->GetPrivateElement();
+    skElementsFrom.SetFormat(Format::COEFFICIENT);
+    auto skElements2 = ckksSKto->GetPrivateElement();
+    skElements2.SetFormat(Format::COEFFICIENT);
+    auto lweskElements = LWEsk->GetElement();
+
+    for (size_t i = 0; i < skElements.GetNumOfElements(); i++) {
+        auto skElementsPlain     = skElements.GetElementAtIndex(i);
+        auto skElementsFromPlain = skElementsFrom.GetElementAtIndex(i);
+        auto skElementsPlainLWE  = skElements2.GetElementAtIndex(i);
+        for (size_t j = 0; j < skElementsPlain.GetLength(); j++) {
+            if (skElementsFromPlain[j] == 0) {
+                skElementsPlain[j] = 0;
+            }
+            else if (skElementsFromPlain[j] == 1) {
+                skElementsPlain[j] = 1;
+            }
+            else
+                skElementsPlain[j] = skElementsPlain.GetModulus() - 1;
+
+            if (j >= lweskElements.GetLength()) {
+                skElementsPlainLWE[j] = 0;
+            }
+            else {
+                if (lweskElements[j] == 0) {
+                    skElementsPlainLWE[j] = 0;
+                }
+                else if (lweskElements[j].ConvertToInt() == 1) {
+                    skElementsPlainLWE[j] = 1;
+                }
+                else
+                    skElementsPlainLWE[j] = skElementsPlain.GetModulus() - 1;
+            }
+        }
+        skElements.SetElementAtIndex(i, skElementsPlain);
+        skElements2.SetElementAtIndex(i, skElementsPlainLWE);
+    }
+
+    skElements.SetFormat(Format::EVALUATION);
+    skElements2.SetFormat(Format::EVALUATION);
+
+    auto ccCKKSto        = ckksSKto->GetCryptoContext();
+    auto oldTranformedSK = ccCKKSto->KeyGen().secretKey;
+    oldTranformedSK->SetPrivateElement(std::move(skElements));
+    auto RLWELWEsk = ccCKKSto->KeyGen().secretKey;
+    RLWELWEsk->SetPrivateElement(std::move(skElements2));
+
+    return ccCKKSto->KeySwitchGen(oldTranformedSK, RLWELWEsk);
+}
+
+std::vector<std::vector<NativeInteger>> ExtractLWEpacked(const Ciphertext<DCRTPoly>& ct) {
+    auto A = ct->GetElements()[1];
+    auto B = ct->GetElements()[0];
+    auto N = B.GetLength();
+
+    auto originalA = A.GetElementAtIndex(0);
+    auto originalB = B.GetElementAtIndex(0);
+    originalA.SetFormat(Format::COEFFICIENT);
+    originalB.SetFormat(Format::COEFFICIENT);
+
+    std::vector<std::vector<NativeInteger>> extracted(2);
+
+    for (uint32_t i = 0; i < N; i++) {
```

Comment:
1. extracted should be resized before it gets more elements (call reserve())
2. can we just copy A and B directly to extracted avoiding the loop?

Reply:
Unfortunately, you can only get down to the NativeVector, not the std::vector<NativeInteger> within NativeVector. If I added a getter method to NativeVector you could do something like:

auto B{(ct->GetElements()[0]).GetElementAtIndex(0)};
B.SetFormat(Format::COEFFICIENT);
auto A{(ct->GetElements()[1]).GetElementAtIndex(0)};
A.SetFormat(Format::COEFFICIENT);
return std::vector<std::vector<NativeInteger>>{B.GetValues().GetVector(), A.GetValues().GetVector()};

Reply:
in any event, you can eliminate extra copies by using

auto B{(ct->GetElements()[0]).GetElementAtIndex(0)};
auto A{(ct->GetElements()[1]).GetElementAtIndex(0)};

Reply:
Done, please check
    extracted[1].reserve(N);
    auto originalAVals = originalA.GetValues();
    extracted[1].insert(extracted[1].end(), &originalAVals[0], &originalAVals[N]);